### ISSUE TITLE
Node Auto Drain

### ISSUE THREAD

**Issue by palma21** [Azure/AKS#2062]

**Comment by djsly** [comment]
I'm curious, what is this exactly ?

**Comment by PaulRiviera** [comment]
I am assuming this has to do with updating the nodes after patching, can someone clarify if this is correct or not?

**Comment by palma21** [comment]
Prototype: https://github.com/juan-lee/nodify

**Comment by DanielHajduk2-TomTom** [comment]
Would love to see this feature being supported by Microsoft directly. Same as for AWS and their https://github.com/aws/aws-node-termination-handler

When can we expect this to be delivered as its already "In Progress" ?

**Comment by MichalKoziorowski-TomTom** [comment]
Hi @palma21. Could you share link to azure documentation telling how to use that GA feature?

**Comment by ivanthelad** [comment]
@heoelri  what we recognized with some simple tests of rebooting a vmss instance 
A instance reboot scheduled via ARM results in Kubernetes raising the following event 
`18m Normal RebootScheduled node/aks-defaultpool-10919361-vmss000000 Node condition RebootScheduled is now: True, reason: RebootScheduled`

The reboot from the ARM perspective appeared to be delayed by around 15 mins. Unsure if aks was blocking this or not.
Pods were not drained from the vm scheduled to be rebooted. 

The questions i have is 
  - Where does this node condition come from 
  - does it block a restart? it seems to
  - why the 15 min delay if we are not asking pods to drain 
@sebader

**Comment by ivanthelad** [comment]
> @heoelri what we recognized with some simple tests of rebooting a vmss instance A instance reboot scheduled via ARM results in Kubernetes raising the following event `18m Normal RebootScheduled node/aks-defaultpool-10919361-vmss000000 Node condition RebootScheduled is now: True, reason: RebootScheduled`
> 
> The reboot from the ARM perspective appeared to be delayed by around 15 mins. Unsure if aks was blocking this or not. Pods were not drained from the vm scheduled to be rebooted.
> 
> The questions i have is
> 
> * Where does this node condition come from
> * does it block a restart? it seems to
> * why the 15 min delay if we are not asking pods to drain
>   @sebader

the 15min delay is by design from Azure. https://docs.microsoft.com/en-us/azure/virtual-machines/linux/scheduled-events#event-scheduling. 

Has AKS  behavior changed with how AKS performs reboots of underlying VMSS instances? 

Previously i thought a reboot would be near instant on a VMSS instance but I maybe wrong here.  From the docs, it states that this 15min delay is activated once the metadata service is first accessed (within 24hours)
https://docs.microsoft.com/en-us/azure/virtual-machines/linux/scheduled-events#enabling-and-disabling-scheduled-events

**Comment by ivanthelad** [comment]
@heoelri   regarding your question 
Why is a scheduled "reboot" not triggering a node autodrain? (based on [this table](https://docs.microsoft.com/en-us/azure/aks/node-auto-repair#node-autodrain))

My guess at this its because of the controller manager has a config pod-eviction-timeout.  which is 5 minutes. 
The default eviction timeout duration is five minutes. from the docs.  if the vm was away more than 5 mins then it makes sense to cordon and drain. w

`If the status of the Ready condition remains Unknown or False for longer than the pod-eviction-timeout (an argument passed to the [kube-controller-manager](https://kubernetes.io/docs/reference/command-line-tools-reference/kube-controller-manager/)), then the [node controller](https://kubernetes.io/docs/concepts/architecture/nodes/#node-controller) triggers [API-initiated eviction](https://kubernetes.io/docs/concepts/scheduling-eviction/api-eviction/) for all Pods assigned to that node. The default eviction timeout duration is five minutes. In some cases when the node is unreachable, the API server is unable to communicate with the kubelet on the node. The decision to delete the pods cannot be communicated to the kubelet until communication with the API server is re-established. In the meantime, the pods that are scheduled for deletion may continue to run on the partitioned node.`
https://kubernetes.io/docs/concepts/architecture/nodes/#condition

Basically a reboot takes around 1 minute. more or less like a like network blip to kubernetes . it would not sense to start cordoning and draining pods because of reboots. if there was more vms been reboot that would result in a load of pods trying to find new homes. 

Compare this to a preempt where the vm will never come back. it makes sense in this case to cordon and drain
